docs/design/variadics.md

```diff
@@ -0,0 +1,644 @@
+-   An `...and` or `...or` expression has type `bool`, and every segment of the
+    operand's type pack must have a type that's convertible to `bool`.
+-   For a `...` tuple element expression or pattern, the segments of the
+    operand's type pack become segments of the type of the enclosing tuple.
+
+### Pattern matching with generalized tuple types
+
+Tuple pattern matching involves unifying the type of the tuple pattern with the
+type of the scrutinee. This becomes substantially more challenging when the
+tuple types can contain variadic segments, because the two tuple types may not
+have the same shape. As a consequence, we don't necessarily know which pattern
+segments each scrutinee segment will match with, or the other way around. For
+example, consider the following code:
+
+```carbon
+fn F(a: i32, ... each b: i32, c: i32);
+
+fn G(... each x: i32) {
+  F(1, 2, ... each x);
+}
+```
+
+If `G` is called with no arguments, the `2` will match with `c`, and otherwise
+the `2` will match with an element of `b`. Similarly, if `G` is called with one
+or more arguments, the last argument will match `c`, and the remaining elements
+(if any) will match values from `b`. However, at type-checking time we don't
+know how many arguments the caller will pass, so we don't know which will occur.
+On the other hand, the `1` will always match `a`.
+
+In general, we want type checking to fail if any possible monomorphization of
+the generic code would fail to typecheck. In this case that means we want type
+checking to fail if any of the potential argument-parameter mappings could fail
+to typecheck after monomorphization. Furthermore, for reasons of readability as
+well as efficiency, we want type checking to fail if any two potential mappings
+would deduce inconsistent values for any deduced parameter. However, in general
+this is intractable, because in the worst case the number of distinct ways to
+map symbolic arguments to parameters is ${2n \choose n}$ for n variadic
+arguments, which is only a factor of $\sqrt{n}$ away from exponential.
+
+Introducing type deduction further complicates the situation. For example:
+
+```carbon
+fn H[each T:! type](a: i32, ... each b: each T, c: String) -> (... each T);
+
+external impl P as ImplicitAs(i32);
+external impl Q as ImplicitAs(String);
+
+fn I(each x: i32, each y: f32, each z: String) {
+  var result: auto = H(... each x, {} as P, ... each y, {} as Q, ... each z);
+}
+```
+
+Here, the deduced type of `result` can have one of four different forms. The
+most general case is `(<i32, |x|>, <P, 1>, <f32, |y|>, <Q, 1>, <String, |z|>)`,
+and the other three cases are formed by omitting the first two and/or last two
+segments (corresponding to the cases where `x` and/or `z` do not match any
+arguments). Extending the type system to support deduction that splits into
+multiple cases would add a fearsome amount of complexity to the type system.
+
+#### Identifying potential matchings
```

Comment:
I'm assuming you would allow multiple arguments to map to the single variadic parameter? This is needed to call a variadic with a non-variadic argument list.

One of the motivating use cases for the current approach was to support the case where functions requiring at least one argument, but the caller puts the extra argument at the end while the callee puts the extra parameter at the beginning (or the other way around).

Reply:
As discussed on Discord and elsewhere, `Min(... each x, y)` is an example of something that wouldn't work with that approach, but your question led me to realize that the rules I had didn't support that case either. However, your approach can be made to work if we extend the type system and syntax to allow `Min` to be written with a single variadic parameter that's required to have at least one element, so that's roughly what I've done here (commit b16a5d9cc7 if you want to see it in isolation from the other comment responses). I'll follow up with changes to p2240.md to discuss the tradeoffs of this approach.

Reply:
I've now updated p2240.md, so this should now be fully ready for review.